src/Data/Singletons/Internal.hs

```diff
@@ -178,6 +222,72 @@ infixl 9 @@
 newtype instance Sing (f :: k1 ~> k2) =
   SLambda { applySing :: forall t. Sing t -> Sing (f @@ t) }
 
+-- | An (explicitly bidirectional) pattern synonym for
+-- defunctionalized singletons (@Sing (f :: k ~> k' ~>
+-- k'')@).
+-- 
+-- As a __constructor__: Same as 'singFun2'. Turns a binary function
+-- on singletons @sTake :: SingFunction2 TakeSym0@ into a
+-- defunctionalized singleton @Sing (TakeSym :: Nat ~> [a] ~> [a])@.
+-- 
+-- @
+-- >>> import Data.Singletons.Prelude.List
+-- >>> :set -XTypeApplications
+-- >>> 
+-- >>> :t SLambda2
+-- SLambda2 :: SingFunction2 f -> Sing f
+-- >>> :t SLambda2 \@TakeSym0
+-- SLambda2 :: SingFunction2 TakeSym0 -> Sing TakeSym0
+-- >>> :t SLambda2 \@TakeSym0 sTake
+-- SLambda2 :: Sing TakeSym0
+-- @
+-- 
+-- This is useful for functions on singletons @sZipWith ::
+-- SingFunction3 ZipWithSym0@ that expect a defunctionalized singleton
+-- as an argument
+-- 
+-- @
+-- sZipWith :: Sing (f::a ~> b ~> c) -> Sing (xs::[a]) -> Sing (ys::[b]) -> Sing (ZipWith f xs ys::[c])
+-- 
+-- sZipWith (SLambda2 \@TakeSym0 sTake) :: Sing (xs::[Nat]) -> Sing (ys::[[a]]) -> Sing (ZipWith TakeSym0 xs ys::[[a]])
+-- @
+-- 
+-- As a __pattern__: Same as 'unSingFun2'. Gets a binary term-level
+-- Haskell function on singletons @Sing (x :: k) -> Sing (y :: k') ->
+-- Sing (f \@\@ x \@\@ y)@ from a defunctionalised @Sing
+-- f@. Alternatively as a record field accessor
+--
+-- @
+-- applySing2 :: Sing (f :: k ~> k' ~> k'') -> SingFunction2 f
+-- @
+pattern SLambda2 :: forall f. SingFunction2 f -> Sing f
+pattern SLambda2 {applySing2} <- (unSingFun2 -> applySing2)
+  where SLambda2 lam2         = singFun2 lam2
```

Comment:
Just for the sake of completeness, I'd add one `COMPLETE` pragma for each of these `SLambda*` pattern synonyms. (They won't ever kick in at the moment due to [Trac #14059](https://ghc.haskell.org/trac/ghc/ticket/14059), but it won't harm anything to add now in anticipation of that bug being fixed one day.)